pallets/rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+                // Deposit an event with the result
+                Self::deposit_event(
+                    RawEvent::FreeCallResult(
+                        sender,
+                        result.map(|_| ()).map_err(|e| e.error),
+                    )
+                );
+            }
+
+            Ok(())
+        }
+    }
+}
+
+impl<T: Config> Module<T> {
+    fn update_account_stats(
+        who: &T::AccountId,
+        window_type: WindowType,
+        stats: &mut ConsumerStats<T::BlockNumber>,
+    ) {
+        stats.current_window_consumed_permits =
+            stats.current_window_consumed_permits.saturating_add(1);
+        StatsByAccount::<T>::insert(who, window_type, stats);
+    }
+
+    /// This function can update stats of a corresponding window,
+    /// if account is eligible to have a free call withing a given window.
+    fn can_account_make_free_call<F>(sender: &T::AccountId, update_stats: F) -> bool
+    where
+        F: FnOnce(&T::AccountId, WindowType, &mut ConsumerStats<T::BlockNumber>) + Copy,
+    {
+        if !T::TrustHandler::is_trusted_account(sender) {
+            return false
+        }
+
+        let current_block = frame_system::Module::<T>::block_number();
+        let windows = T::RateConfigs::get();
+        let mut has_free_calls = false;
+
+        for (i, window) in windows.into_iter().enumerate() {
+            let window_type = i as WindowType;
+
+            // Calculate the current window
+            let current_window = current_block / window.period;
+
+            // Calculate the number of blocks elapsed in the current window.
+            let current_window_elapsed = current_block - current_window * window.period;
+
+            let reset_stats = |previous_window_consumed_permits| {
+                ConsumerStats::new(current_window, previous_window_consumed_permits)
+            };
+
+            // Get stats for this type of window
+            let mut stats =
+                StatsByAccount::<T>::get(&sender, window_type).unwrap_or_else(|| reset_stats(0));
+
+            // If no free premits were used in the preceding window set both
+            // current_window_consumed_permits and previous_window_consumed_permits to 0, otherwise
+            // set current_window_consumed_permits to 0 and previous_window_consumed_permits to
+            // current_window_consumed_permits
+            if stats.last_window == current_window - 1u32.into() {
+                stats = reset_stats(stats.current_window_consumed_permits);
+            } else if stats.last_window < current_window {
+                stats = reset_stats(0);
+            }
+
+            // Calculate the number of permits consumed using sliding window rate limiting
+            // algorithm.
+            let permits_used = ((window.period - current_window_elapsed) *
+                stats.previous_window_consumed_permits.into() /
+                window.period) +
+                stats.current_window_consumed_permits.into();
```

Comment:
Consider splitting into more primitive calculations. Furthermore, `Saturating` is good to use here.